marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.smart-mastering/matcher-impl/matcher-impl.xqy

```diff
@@ -361,6 +375,55 @@ declare function match-impl:build-boost-query(
   ))
 };
 
+(:
+ : Build a series of queries that match items with present but
+ : different values for declared-unique fields.
+ : @param $document  the source document from which property values are drawn
+ : @param $is-json  true if source document is JSON
+ : @param $values-by-qname  map:map that organizes document values by QName
+ : @param $match-options  XML or JSON representation of match options
+ : @return a series of cts:and-query items
+ :)
+declare function match-impl:build-uniqueness-queries(
+  $document as node(),
+  $is-json as xs:boolean,
+  $values-by-qname as map:map,
+  $match-options as item()
+)
+{
+  let $options :=
+    if ($match-options instance of object-node()) then
+      opt-impl:options-from-json($match-options)
+    else
+      $match-options
+  for $property-def in $options/matcher:property-defs/matcher:property
+    return if (xs:boolean($property-def/matcher:unique)) then
```

Comment:
I like the idea of configuring this via a property, but it hurts my head to think through all the negatives here - i.e. this returns a query that is included in an or-query that is part of an and-not-query, and the returned query is an and-query that includes a not-query around an or-query. 

Abstracting up, it seems simpler and sufficient to construct the overall match query to be an AND query like this:

- An OR query on all the match actions
- (Optional) an AND query on what could be called "filter" or "inclusion" queries

That can then be tossed into a cts:and-not-query, with the not-query being as it is today. 

The "inclusion" query can then be expressed as an OR query (using a concrete example of SSN):

- Either the matching document has the same SSN
- Or it has an SSN element/property with no value
- Or it does not have an SSN element/property

This query wouldn't have a weight on it because weighting doesn't apply here - this is a hard and fast rule that all matches must meet. 

So the way I'd organize this is by creating a function to build up zero or more "inclusion" queries. One source of those could be the presence of the matcher:unique element.

Reply:
I agree especially with the comments on the double NOT logic. It should be pretty simple to avoid that by using the existing `$filter-query` parameter and extending with additional queries if the unique option is set on any of the values.